src/ZPublisher/utils.py

```diff
@@ -100,3 +101,119 @@ def basic_auth_decode(token):
+       or worse: a combination of strings and bytes.
+    2. Replace deprecated ulines, utext, utoken, and ustring properties
+       with their non-unicode variant, using native strings.
+
+    See https://github.com/zopefoundation/Zope/issues/987
+
+    Since Zope 5.3, a lines property stores strings instead of bytes.
+    But there is no migration yet.  (We do that here.)
+    Result is that getProperty on an already created lines property
+    will return the old value with bytes, but a newly created lines property
+    will return strings.  And you might get combinations.
+
+    Also since Zope 5.3, the ulines property type is deprecated.
+    You should use a lines property instead.
+    Same for a few others: utext, utoken, ustring.
+    The unicode variants are planned to be removed in Zope 6.
+
+    Intended usage:
+    app.ZopeFindAndApply(app, apply_func=fix_properties)
+    """
+    if path is None:
+        # When using ZopeFindAndApply, path is always given.
+        # But we may be called by other code.
+        if hasattr(object, 'getPhysicalPath'):
+            path = '/'.join(object.getPhysicalPath())
+        else:
+            # Some simple object, for example in tests.
+            # We don't care about the path then, it is only shown in logs.
+            path = "/dummy"
+
+    try:
+        prop_map = obj.propertyMap()
+    except AttributeError:
+        # Object does not inherit from PropertyManager.
+        # For example 'MountedObject'.
+        return
+
+    for prop_info in prop_map:
+        # Example: {'id': 'title', 'type': 'string', 'mode': 'w'}
+        prop_id = prop_info.get("id")
+        current = obj.getProperty(prop_id)
+        if current is None:
+            continue
+        new_type = prop_type = prop_info.get("type")
+        if prop_type == "lines":
+            new = _string_tuple(current)
+        elif prop_type == "ulines":
+            new_type = "lines"
+            new = _string_tuple(current)
+        elif prop_type == "utokens":
+            new_type = "tokens"
+            new = _string_tuple(current)
+        elif prop_type == "utext":
+            new_type = "text"
+            new = safe_unicode(current)
+        elif prop_type == "ustring":
+            new_type = "string"
+            new = safe_unicode(current)
+        else:
+            continue
+        if prop_type != new_type:
+            # Replace with non-unicode variant.
+            # This could easily lead to:
+            # Exceptions.BadRequest: Invalid or duplicate property id.
+            #   obj._delProperty(prop_id)
+            #   obj._setProperty(prop_id, new, new_type)
+            # So fix it by using internal details.
+            for prop in obj._properties:
+                if prop.get("id") == prop_id:
+                    prop["type"] = new_type
+                    # This is a tuple, so force a safe, just to be sure.
+                    obj._properties = obj._properties
```

Comment:
I neither understand the comment (is there something missing after the `safe`?, nor do I understand the assignment expression.

Reply:
We change a dict in a tuple. Without this assignment, which marks the object as changed, the change may not be persisted in the zodb.

Reply:
> Without this assignment, which marks the object as changed, the change may not be persisted in the zodb.

If `obj` is a persistent object, isn't the correct thing to do to set `obj._p_changed = True`? If it may or may not be persistent, then `if hasattr(obj, '_p_changed'): obj._p_changed = True` works if you don't want to do `if IPersistent.providedBy(obj):`.

Reply:
I know `obj._p_changed = True` would work, and am pretty sure we only get persistent objects in here, but to me it looks more magical than the current code. It's debatable though.
Do you think the current code is okay with the updated comment?

Reply:
To me, setting `obj._p_changed = True` is much more explicit and self explanatory than `obj.thing = obj.thing`. Setting `_p_changed` would probably eliminate the need for a comment at all.

The former (`_p_changed`) is a documented part of [the `IPersistent` protocol](https://persistent.readthedocs.io/en/latest/api/interfaces.html#persistent.interfaces.IPersistent); the latter relies on a side effect (and the side effect may not work depending on the implementation of the object, for example if the class has made `thing` a special type of descriptor).

Reply:
Okay, you convinced me. Done.